b2g/components/RemoteControlService.jsm

```diff
@@ -0,0 +1,395 @@
+      var conn = new Connection(input, output, this, socket.port, trans.port, connectionNumber);
+      let handler = new EventHandler(conn);
+
+      input.asyncWait(handler, 0, 0, Services.tm.mainThread);
+    } catch (e) {
+      DEBUG && debug("Error in initial connection: " + e);
+      trans.close(Cr.NS_BINDING_ABORTED);
+      return;
+    }
+
+    this._connections.set(connectionNumber, conn);
+    DEBUG && debug("Start connection " + connectionNumber);
+  },
+
+  // Close all connection when socket closed
+  onStopListening: function(socket, status) {
+    DEbug && debug("Shut down server on port " + socket.port);
+
+    this._connections.forEach(function(value, key){
+      let connection = value;
+      connection.close();
+    });
+
+    this._socketClosed = true;
+  },
+
+  // PRIVATE FUNCTIONS
+  _doStart: function(aResolve, aReject) {
+    DEBUG && debug("doStart");
+
+    if (this._socket) {
+      aReject("SocketAlreadyInit");
+      return;
+    }
+
+    this._doQuit = this._socketClosed = false;
+
+    // Monitor xpcom-shutdown to stop service and clean up
+    Services.obs.addObserver(this, "xpcom-shutdown", false);
+
+    // Start TLSSocketServer with self-signed certification
+    Cc["@mozilla.org/psm;1"].getService(Ci.nsISupports);
+    certService.getOrCreateCert("RemoteControlService", {
+      handleCert: function(cert, result) {
+        if(result) {
+          aReject("getOrCreateCert " + result);
+        } else {
+          let self = RemoteControlService;
+
+          try {
+            // Try to get random port
+            let ios = Cc["@mozilla.org/network/io-service;1"].getService(Ci.nsIIOService);
+            let socket;
+            for (let i = 100; i; i--) {
+              let temp = Cc["@mozilla.org/network/tls-server-socket;1"].createInstance(Ci.nsITLSServerSocket);
+              temp.init(self._port, false, MAX_CLIENT_CONNECTIONS);
+              temp.serverCert = cert;
+
+              let allowed = ios.allowPort(temp.port, "tls");
+              if (!allowed) {
+                DEBUG && debug("Warning: obtained TLSServerSocket listens on a blocked port: " + temp.port);
+              }
+
+              if (!allowed && self._port == -1) {
+                DEBUG && debug("Throw away TLSServerSocket with bad port.");
+                temp.close();
+                continue;
+              }
+
+              socket = temp;
+              break;
+            }
+
+            if (!socket) {
+              throw new Error("No socket server available. Are there no available ports?");
+            }
+
+            DEBUG && debug("Listen on port " + socket.port + ", " + MAX_CLIENT_CONNECTIONS + " pending connections");
+
+            socket.serverCert = cert;
+            socket.setSessionCache(false);
+            socket.setSessionTickets(false);
+            socket.setRequestClientCertificate(Ci.nsITLSServerSocket.REQUEST_NEVER);
+
+            socket.asyncListen(self);
+            self._port = socket.port;
```

Comment:
In this patch you'll reuse the same port after restarting remote control service, is this the expected behavior? Is there any reason we cannot always use random port after restarting service?